### DISCUSSION TITLE
Compatibility with ESP32-BLE-Gamepad

### DISCUSSION THREAD

**Discussion by tawyy** [darthcloud/BlueRetro#311] [Ideas]
Hello,
I am very interested with this project and i plan to make all my console with officiel controller wireless.
In order to do that, i wanted to make custom PCB (exactly like the 8BitDo Mod Kit) for all of my needed controllers with the [ESP32-BLE-Gamepad](https://github.com/lemmingDev/ESP32-BLE-Gamepad) 
I am actually working on a Dreamcast Version, Neo Geo (All Arcade Stick compatible with screw terminals), Sega Saturn and maybe Xbox 360 (i have a lot of xbox 360 game controller shells lying in my stuff, and cheap xbox 360 clone controller are affordable).

Thank you

**Comment by darthcloud** [comment]
See #293

This project as-is is not a representative examples of how a a typical commercial HID controller work.

As a tell everyone edit the project to look like an Xbox X|S controller and it will work. #293 got pointer on how to do so.

**Reply by tawyy** [reply]
If the blueretro can sync with it i'm ok.
I needed to adapt the code because of the triggers on the Dreamcast controller, they work with a hall effect sensor, is working on active high and the sensivity seems to be different of a regular 10k potentiometer. I dont know if i can invert the behavior of the circuit. i will try it ;)

**Reply by darthcloud** [reply]
You should talk with @GamingNJncos he already got it working:
https://twitter.com/GamingNJncos/status/1530609330986287104?s=20&t=eZCUw7VqLtuxkz28whDfSA

**Comment by GamingNJncos** [comment]
Sync works fine.  Axis are an issue though.  Notes on fix in link below. Just use map() to fix or invert to blegamepad values for triggers or axis as needed.


 https://github.com/lemmingDev/ESP32-BLE-Gamepad/issues/104

**Reply by darthcloud** [reply]
\<rant\>
I still think the issue is on the ESP32-BLE-Gamepad side, IMHO they should target to make their project behave like commercial gamepad.
While technically per HID spec nothing is stopping you from making your axes signed, I have yet to see a commercial gamepad doing so. It's always unsigned AFAICT in pratice.
\</rant\>

~~v1.6.2_hid_sign_axes_fix.zip~~

Anyway, here a build that should potentially fix the signed axes issue in BlueRetro.
@GamingNJncos can you test it? (Need a version of your controller FW that use signed axes).
I got no quick way to test it myself.

**Reply by GamingNJncos** [reply]
Of course, no problem. 

Will give this a shot later tonight and report back. Thanks!

**Reply by tawyy** [reply]
Yes, I try to invert the values on the map function and it works well with the potentiometer. For the Hall effect sensor I just have to change the sensitivity on the map function, not a big deal :)
I have weird issue on my test breadboard. When I put a pot on gpio34 and configure gpio35 as standard button (input pullup, active low), when gpio 34 reach half of the value on joy.cpl, it activate gpio 35 :( I had to move it to another gpio to avoid this.

**Reply by GamingNJncos** [reply]
Running through initial testing, this appears to work great. I'll keep poking around and let you know if I find anything odd. To be specific, I simply used OTA to upload the saturn bin and after a reboot (and recalling to switch back to gamepad Alt Mode) just works on older versions of my code with the full advertised range.

I can understand and entirely agree a 64k range for something like an stick or a trigger is pretty silly.  There are some other really great use cases with the combo of BLEGamepad + Blueretro you may or may not have considered and the doors it opens up. Once again, appreciate the support.

**Reply by tawyy** [reply]
And of course, thank you for the support ;)
I am working on a Dreamcast PCB to make a wireless Esp32 powered controller. I think I have everything to do a schematic. I will put a charge/discharge 5V lipo module already mounted, an ams1117 3.3V and the esp32 Wroom. My only issue is the limitation of Eagle PCB design right now. The free version is limited to a 10cm2 and a Dreamcast controller pcb is bigger… I need time to work on it. I am waiting for your project of internal blueretro install :)

**Reply by darthcloud** [reply]
Well being sign vs unsign don't change the range, 16 bit with either will give 64k range in both directions, it just change the min, middle, max value, -32k, 0, 32k for sign vs 0, 32k, 64k for unsigned.

BlueRetro should support 32bit axes btw but untested. But anything over that like 64bit will not be supported.

**Reply by darthcloud** [reply]
Looking at it again the triggers maximum looked wrong here another build:
~~v1.6.2_hid_sign_axes_fix2.zip~~

**Comment by darthcloud** [comment]
~~v1.6.2_hid_sign_axes_fix3.zip~~

That should be my final iteration

**Comment by darthcloud** [comment]
v1.6.3 include theses changes:

https://github.com/darthcloud/BlueRetro/releases/tag/v1.6.3

**Reply by GamingNJncos** [reply]
Thank you once again!!
